src/mapping.rs

```diff
@@ -130,6 +136,47 @@ impl<'s> MappingSummary<'s> {
     }
 }
 
+/// A class index for a proguard mapping file.
+///
+/// This is fundamentally a map from class names to offset ranges. The value
+/// of `class` in this map is the range range of bytes between which the
+/// mapping information for `class` is found in the mapping file
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ClassIndex<'a>(BTreeMap<Cow<'a, str>, Range<usize>>);
+
+impl<'a> ClassIndex<'a> {
+    /// Gets the range information for a class from this index.
+    pub fn get(&self, class: &str) -> Option<Range<usize>> {
+        self.0.get(class).cloned()
+    }
+
+    /// Reads a class index from its text representation (see [`write`](Self::write)).
+    pub fn parse(source: &[u8]) -> Option<Self> {
+        let mut inner = BTreeMap::new();
+        for line in source.lines() {
+            let line = line.ok()?;
+            let (name, rest) = line.split_once(':')?;
+            let (start, end) = rest.split_once(':')?;
+            let (start, end) = (start.parse().ok()?, end.parse().ok()?);
+            inner.insert(Cow::Owned(name.into()), start..end);
```

Comment:
```suggestion
    pub fn parse(source: &'a [u8]) -> Option<Self> {
        let mut inner = BTreeMap::new();
        for line in source.lines() {
            let line = line.ok()?;
            let (name, rest) = line.split_once(':')?;
            let (start, end) = rest.split_once(':')?;
            let (start, end) = (start.parse().ok()?, end.parse().ok()?);
            inner.insert(Cow::Borrowed(name), start..end);
```

I think this should work?

hm, but okay, maybe there is a difference between using the proguard buffer lifetime, or the index buffer lifetime?

Reply:
Unfortunately it doesn't work as written. The reason is that `lines` (from the `BufRead` trait) is an iterator over `String`, not `&str`. AFAICT there's no way to iterate over the lines of a `&[u8]` in a borrowing fashion. We could make this function take `&str` instead, that would fix it. The up-front UTF8 validation probably wouldn't be too big of a problem because these indices are small, and it has to be done at some point anyway because we're reading strings.